source-code/core/messages/src/variant.ts

```diff
@@ -0,0 +1,242 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+import type { LanguageTag } from "@inlang/language-tag"
+import type { Message, Variant } from "./schema.js"
+import type { Result } from "@inlang/result"
+import {
+	MessagePatternsForLanguageTagDoNotExistError,
+	MessageVariantAlreadyExistsError,
+	MessageVariantDoesNotExistError,
+} from "./errors.js"
+
+/**
+ * Get the variant of a message
+ *
+ * All actions are immutable.
+ *
+ * @example
+ * 	const variant = getVariant(message, { where: { languageTag: "en", selectors: { gender: "male" }}});
+ */
+export function getVariant(
+	message: Message,
+	args: {
+		where: {
+			languageTag: LanguageTag
+			selectors?: Variant["match"]
+		}
+	},
+): Variant | undefined {
+	const variant = matchMostSpecificVariant(message, args.where.languageTag, args.where.selectors)
+	if (variant) {
+		//! do not return a reference to the message in a resource
+		//! modifications to the returned message will leak into the
+		//! resource which is considered to be immutable.
+		return structuredClone(variant)
```

Comment:
Do we need to clone the object? We already do that when querying messages.

Reply:
I think we want the variant to me immutable right? The same way as we use the query utility functions in the code base before the refactoring